### PR TITLE
Adding new module vg/deconstruct

### LINKED ISSUE #1135: new module: vg/deconstruct
I think it would be good to have a module for [vg](https://bioconda.github.io/recipes/vg/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

https://github.com/nf-core/pangenome/issues/70

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #1135 

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by heuermh** [comment]
Will require proper test data, see https://github.com/nf-core/pangenome/issues/74

**Comment by subwaystation** [comment]
https://github.com/nf-core/test-datasets/pull/406 should solve this problem ;)

**Comment by heuermh** [comment]
Fails with
```
 error [vg deconstruct]: Reference path "pangenome.fa" not found in graph/gbwt
```

File: modules/vg/deconstruct/main.nf [hunk 1]
```diff
@@ -0,0 +1,35 @@
+process VG_DECONSTRUCT {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda (params.enable_conda ? 'bioconda::vg=1.38.0' : null)
```
Comment (subwaystation) [inline modules/vg/deconstruct/main.nf:5]: @heuermh Could you please update to v1.43.0? https://anaconda.org/bioconda/vg.

File: tests/modules/vg/deconstruct/main.nf [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { VG_DECONSTRUCT } from '../../../../modules/vg/deconstruct/main.nf'
+
+workflow test_vg_deconstruct {
+    input = [ [ id:'test' ], // meta map
+              [ file(params.test_data['homo_sapiens']['pangenome']['pangenome_fa'], checkIfExists: true)],
```
Comment (subwaystation) [inline tests/modules/vg/deconstruct/main.nf:9]: I think you misunderstood `vg deconstruct`. `--path` takes one or several path names in the graph as a string, not a FASTA input.

File: modules/vg/deconstruct/main.nf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    vg deconstruct \\
+        --threads $task.cpus \\
+        --path=$fasta \\
```
Comment (subwaystation) [inline modules/vg/deconstruct/main.nf:26]: This must be a string containing a list of comma-separated paths in the graph. Possible is also one path.

File: tests/modules/vg/deconstruct/main.nf [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { VG_DECONSTRUCT } from '../../../../modules/vg/deconstruct/main.nf'
+
+workflow test_vg_deconstruct {
```
[the remark this answers is not in the document]
Comment (subwaystation) [reply tests/modules/vg/deconstruct/main.nf:7]: I just realized you won't need the FASTA. See my other comments.

**Comment by subwaystation** [comment]
@nvnieuwk Why are no tests ran?

**Comment by subwaystation** [comment]
image

I would like to have an optional input files. Is that possible somehow? Or should they be specified via `args`? I fear they are no Channel, then, right? Or can I also hand over a Channel?

**Comment by subwaystation** [comment]
> @heuermh example of optional input:
> 
> https://github.com/nf-core/modules/blob/3c254794f9850b9588b0e5435cabec12e6995121/modules/nf-core/adapterremoval/main.nf#L29
> 
> where `adapterlist` is from the inputchannel

@jfy133 And the optional input Channel can be an empty Channel then? EMPTY == 0 in the if condition?

**Comment by nvnieuwk** [comment]
> > @heuermh example of optional input:
> > https://github.com/nf-core/modules/blob/3c254794f9850b9588b0e5435cabec12e6995121/modules/nf-core/adapterremoval/main.nf#L29
> > 
> > where `adapterlist` is from the inputchannel
> 
> @jfy133 And the optional input Channel can be an empty Channel then? EMPTY == 0 in the if condition?

If a channel is empty the check will automatically return `false` so no need to compare :)

File: modules/nf-core/vg/deconstruct/main.nf [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    conda (params.enable_conda ? 'bioconda::vg=1.43.0' : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/vg:1.43.0--h9ee0642_0' :
+        'quay.io/biocontainers/vg:1.43.0--h9ee0642_0' }"
+
+    input:
+    tuple val(meta), path(gfa)
+
+    output:
+    tuple val(meta), path("*.vcf"), emit: vcf
```
[the remark this answers is not in the document]
Comment (subwaystation) [reply modules/nf-core/vg/deconstruct/main.nf:14]: It can only emit plain VCF files.

**Comment by subwaystation** [comment]
@nvnieuwk The test is not running through anymore, but I have no idea how to fix it. Because Nextflow is just aborting it's missing without any kind of explanation.

**Comment by subwaystation** [comment]
Got it going. But what is the difference between `Channel.empty()` and `[]`?

**Comment by nvnieuwk** [comment]
> Got it going. But what is the difference between `Channel.empty()` and `[]`?

I'm not entirely sure myself but I think `Channel.empty()` returns an empty channel while `[]` is just an empty array which corresponds to an empty `value` channel. But you should probably ask someone else too about this :grin:

**Comment by jfy133** [comment]
> > Got it going. But what is the difference between `Channel.empty()` and `[]`?
> 
> I'm not entirely sure myself but I think `Channel.empty()` returns an empty channel while `[]` is just an empty array which corresponds to an empty `value` channel. But you should probably ask someone else too about this grin

The answer :point_up:

**Comment by subwaystation** [comment]
@heuermh This is good to go now from my point of view.

**Review by subwaystation — APPROVED** [review]
LGTM!

**Review by nvnieuwk — APPROVED** [review]

**Comment by heuermh** [comment]
Thank you, @subwaystation @nvnieuwk @jfy133!

**Pull request merged**